go/vt/vttablet/tabletserver/tabletserver.go

```diff
@@ -237,6 +239,35 @@ func (tsv *TabletServer) loadQueryTimeout() time.Duration {
 	return time.Duration(tsv.QueryTimeout.Load())
 }
 
+func (tsv *TabletServer) loadQueryTimeoutWithTx(txID int64, options *querypb.ExecuteOptions) time.Duration {
+	timeout := tsv.loadQueryTimeoutWithOptions(options)
+
+	if txID == 0 {
+		return timeout
+	}
+
+	// fetch the transaction timeout.
+	txTimeout := tsv.config.TxTimeoutForWorkload(querypb.ExecuteOptions_OLTP)
+
+	// Use the smaller of the two values (0 means infinity).
+	return smallerTimeout(timeout, txTimeout)
+}
+
+func (tsv *TabletServer) loadQueryTimeoutWithOptions(options *querypb.ExecuteOptions) time.Duration {
```

Comment:
`loadQueryTimeoutWithOptions` and `loadAuthoritativeTimeout` could be simpler as a single function, because `loadAuthoritativeTimeout` isn't used anywhere else.

```go
func (tsv *TabletServer) loadQueryTimeoutWithOptions(options *querypb.ExecuteOptions) time.Duration {
	if options == nil || options.Timeout == nil {
		return time.Duration(tsv.QueryTimeout.Load())
	}
	return time.Duration(options.GetAuthoritativeTimeout()) * time.Millisecond
}
```